### DISCUSSION TITLE
Cors issue

### DISCUSSION THREAD

**Discussion by fv316** [ory/kratos#986] [Q&A] [answered]
Hey guys, I am currently trying to get the Kratos registration flow working on a react app that I am creating. I have borrowed some inspiration from the react-native app and the node app that you guys already created. However, I am having CORS issues with the completeSelfServiceRegistrationFlowWithPasswordMethod call when using the @oryd/kratos-client JS package (I have also tried to use fetch directly). I have tried running everything locally (the react app on 127.0.0.1:3000 and kratos on 127.0.0.1:4434) and also on our domain (http://kratos.example.com and http://react.example.com). Get requests are working but the preflight request is failing for post requests:
```
Access to XMLHttpRequest at 'http://127.0.0.1:3000/error?error=ad779544-b91e-4837-9246-471ae0c43fa7' (redirected from 'http://127.0.0.1:4433/self-service/registration/methods/password?flow=1e82763e-f9e8-45f0-aa33-4db26ac8c6a5') from origin 'http://127.0.0.1:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```
My kratos config looks something like this (but I have tried many other options)
```
cors:
  allowed_origins:
    - http://127.0.0.1:3000 # I have tried with "*"

  enabled: true
```
The request payload being sent looks like this:
```
{
    "csrf_token" : "qweqwe...",
    "password": "qweqweqwe",
    "traits.email" : "qwe@qwe.qwe",
    "traits.name.first" : "qwe",
    "traits.name.last": "qwe"
}
```

**Comment by vinckr** [comment]
Hey fv316. 
I am sure you already saw the [setup page for CORS on Kratos](https://www.ory.sh/kratos/docs/guides/setting-up-cors/). 
But have you also checked [Common Cookie and CSRF Pitfalls](https://www.ory.sh/kratos/docs/debug/csrf/) ?


` No 'Access-Control-Allow-Origin' header is present on the requested resource.`
Did you set allowed headers in the config?
```
      allowed_headers:
        - Authorization
        - Cookie
      exposed_headers:
        - Content-Type
        - Set-Cookie
```

**Reply by fv316** [reply]
```kratos:
  development: true

  autoMigrate: true

  config:
    
    version: v0.5.5-alpha.1

    dsn: "psql_url"

    serve:
      public:
        base_url: http://kratos.user.example.com
        cors:
          enabled: true
          allowed_origins:
            - "http://react.ui.example.com"
          allowed_methods:
            - POST
            - GET
            - PUT
            - PATCH
            - DELETE
          allowed_headers:
            - Authorization
            - Cookie
          exposed_headers:
            - Content-Type
            - Set-Cookie
          debug: true

      admin:
        base_url: https://kratos.admin.example.com

    selfservice:
      default_browser_return_url: http://react.ui.example.com/
      whitelisted_return_urls:
        - http://react.ui.example.com

      methods:
        password:
          enabled: true

      flows:
        error:
          ui_url: http://react.ui.example.com/error

        settings:
          ui_url: http://react.ui.example.com/settings
          privileged_session_max_age: 15m

        recovery:
          enabled: false

        verification:
          enabled: false
          
        logout:
          after:
            default_browser_return_url: http://react.ui.example.com/auth/login

        registration:
          lifespan: 1m
          after:
            password:
              hooks:
                - hook: session
              default_browser_return_url: http://react.ui.example.com/dashboard
            oidc:
              hooks:
                - hook: session
              default_browser_return_url: http://react.ui.example.com/dashboard
            default_browser_return_url: http://react.ui.example.com/dashboard
          ui_url: http://react.ui.example.com/auth/registration

        login:
          lifespan: 10m
          after:
            password:
              hooks:
                - hook: revoke_active_sessions
              default_browser_return_url: http://react.ui.example.com/dashboard
            oidc:
              hooks:
                - hook: revoke_active_sessions
              default_browser_return_url: http://react.ui.example.com/dashboard
            default_browser_return_url: http://react.ui.example.com/dashboard
          ui_url: http://react.ui.example.com/auth/login

    log:
      level: debug
      format: text
      leak_sensitive_values: true

    secrets:
      cookie:
        - PLEASE-CHANGE-ME-I-AM-VERY-INSECURE

    hashers:
      argon2:
        parallelism: 1
        memory: 131072
        iterations: 2
        salt_length: 16
        key_length: 16

    identity:
      default_schema_url: "some_url"
      schemas:
        - id: customer
          url: "some_url"

    courier:
      smtp:
        connection_uri: smtps://test:test@mailslurper:1025/?skip_ssl_verify=true

    session:
      cookie:
        persistent: false
        path: '/'
        same_site: Lax
        domain: 'example.com'
      lifespan: 1h
```

**Reply by fv316** [reply]
I have tried all those solutions, my current kratos.yml files looks like this. I tried the same thing on localhost by replacing the appropriate urls.

**Reply by fv316** [reply]
Using this config the NODE UI application works but my react application does not. This leads me to believe that it is some client side problem of the request not being sent correctly. Is there an example payload of what should be sent to the complete registration flow endpoint? I could not find one on the REST API docs.

**Reply by aeneasr** [reply]
Hello, it looks like you are using Kubernetes. Are you sure that the Ingress is not stripping the CORS headers, or rather, that the ingress is incorrectly replying with the wrong CORS headers? Your configuration should work, and if exposed locally (without a proxy in front) should yield the correct CORS headers. This is tested and confirmed working.

You should also check the logs to see if the requests actually make it to ORY Kratos or if they do not (my gut feeling says they do not).

**Reply by aeneasr** [reply]
Ok, I just checked again. It appears that you are using AJAX to submit the form. This is not supported right now. You need to regularly submit - like an HTML form. AJAX will be better supported starting 0.6

**Reply by fv316** [reply]
What do you mean by submit the form regularly? I have tried using simple post requests using fetch instead of the client API but I get the same issue

**Reply by aeneasr** [reply]
Just submit the form. Don't use AJAX. Don't use fetch. Let the user click the submit button **and don't do anything** - a regular HTML form submission.

**Comment by vinckr** [comment, marked as answer]
Since the answer was a bit buried in the thread:

>  You are using AJAX to submit the form. This is not supported right now. You need to regularly submit - like an HTML form. AJAX will be better supported starting 0.6

**Reply by aaadilmaq0** [reply]
@vinckr I am running into the same issue, with the only difference being that I am using Angular. I have including withCredentials: true in all my requests and also I have submitted like an HTML form but it keeps giving me CORS error with the following message: the value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'.

My UI is running on 127.0.0.1:4455 and kratos on 127.0.0.1:4433. This is my kratos.yml

`version: v0.4.6-alpha.1

dsn: memory

serve:
  public:
    base_url: http://127.0.0.1:4433/
    cors:
      enabled: true
      allowed_origins:
        - "http://127.0.0.1:4455"
      allowed_methods:
        - POST
        - GET
        - PUT
        - PATCH
        - DELETE
      allowed_headers:
        - Authorization
        - Content-Type
      exposed_headers:
        - Content-Type
        - Set-Cookie
      allow_credentials: true
      debug: true
  admin:
    base_url: http://kratos:4434/

selfservice:
  default_browser_return_url: http://127.0.0.1:4455/
  whitelisted_return_urls:
    - http://127.0.0.1:4455

  methods:
    password:
      enabled: true

  flows:
    error:
      ui_url: http://127.0.0.1:4455/error

    settings:
      ui_url: http://127.0.0.1:4455/settings
      privileged_session_max_age: 15m

    recovery:
      enabled: true
      ui_url: http://127.0.0.1:4455/recovery

    verification:
      enabled: true
      ui_url: http://127.0.0.1:4455/verify
      after:
        default_browser_return_url: http://127.0.0.1:4455/

    logout:
      after:
        default_browser_return_url: http://127.0.0.1:4455/auth/login

    login:
      ui_url: http://127.0.0.1:4455/auth/login
      lifespan: 10m

    registration:
      lifespan: 10m
      ui_url: http://127.0.0.1:4455/auth/registration
      after:
        password:
          hooks:
            -
              hook: session

log:
  level: debug
  format: text
  leak_sensitive_values: true

secrets:
  cookie:
    - PLEASE-CHANGE-ME-I-AM-VERY-INSECURE

hashers:
  argon2:
    parallelism: 1
    memory: 131072
    iterations: 2
    salt_length: 16
    key_length: 16

identity:
  default_schema_url: file:///etc/config/kratos/identity.schema.json

courier:
  smtp:
    connection_uri: smtps://test:test@mailslurper:1025/?skip_ssl_verify=true
`

**Reply by vinckr** [reply]
Hm what is the response by Ory Kratos? You can curl for the complete response, for example:
`curl -s "http://127.0.0.1:4434/self-service/login/flows?id=<flow_id>" \ | jq `

Also can you try changing the URLs in your schema from 
http://127.0.0.1:4433/
to 
http://127.0.0.1:4433
maybe that is it. So omit the final "/".

**Reply by vinckr** [reply]
If that does not solve your problem, could you open a separate discussion for this please?

**Reply by aaadilmaq0** [reply]
Ok I solved this issue but now another issue is coming which I will discuss in a separate issue.

**Reply by jay-kanakala** [reply]
Bro, how did you solve it.. pls tell me.. i am also facing same issue

**Reply by vinckr** [reply]
@jay-kanakala 
What exactly is your issue?

**Reply by jay-kanakala** [reply]
Well, I am facing few issues mainly not being able to send verification and recovery mails through API flow because it keeps giving CSRF missing bad request error. I guess I have to wait for 0.6 release. For now I have to look for other alternative

**Comment by aaadilmaq0** [comment]
I weren't able to solve this issue.

**Comment by vinckr** [comment]
By the way I found a pretty good article on CORS:
https://ieftimov.com/post/deep-dive-cors-history-how-it-works-best-practices/
Maybe it can be of some help.
